oauth2-useragent-core/src/main/java/com/microsoft/alm/oauth2/useragent/UserAgentImpl.java

```diff
@@ -157,70 +205,69 @@ void addClassPathToCommand(final List<String> classPath, final List<String> comm
         command.add("-classpath");
         //noinspection ToArrayCallWithZeroLengthArrayArgument
         final String[] classPathComponents = classPath.toArray(EMPTY_STRING_ARRAY);
-        final String classPathString = join(pathSeparator, classPathComponents);
+        final String classPathString = StringHelper.join(pathSeparator, classPathComponents);
         command.add(classPathString);
     }
 
-    static String join(final String separator, final String[] value)
-    {
-        if (value == null)
-            throw new IllegalArgumentException("value is null");
-
-        // "If separator is null, an empty string (String.Empty) is used instead."
-        final String sep = separator == null ? "" : separator;
+    static void throwUnsupported(final Map<Provider, List<String>> unmetRequirements) {
+        final StringBuilder sb = new StringBuilder("I don't support your platform yet.");
+        describeUnmetRequirements(unmetRequirements, sb);
+        sb.append(NEW_LINE);
+        sb.append("Please send details about your operating system version, Java version, 32- vs. 64-bit, etc.");
+        sb.append(NEW_LINE);
+        sb.append("The following System Properties and Environment Variables would be very useful.");
+        sb.append(NEW_LINE);
 
-        final StringBuilder result = new StringBuilder();
+        final Properties properties = System.getProperties();
+        appendProperties(properties, sb);
+        sb.append(NEW_LINE);
 
-        if (value.length > 0) {
-            result.append(value[0] == null ? "" : value[0]);
-            for (int i = 1; i < value.length; i++) {
-                result.append(sep);
-                result.append(value[i] == null ? "" : value[i]);
-            }
-        }
+        final Map<String, String> variables = System.getenv();
+        appendVariables(variables, sb);
 
-        return result.toString();
+        throw new IllegalStateException(sb.toString());
     }
 
-    static Provider determineProvider(final String userAgentProvider) {
-        return determineProvider(userAgentProvider, Provider.PROVIDERS);
-    }
+    static Provider scanProviders(final String userAgentProvider, final List<Provider> providers, final Map<Provider, List<String>> destinationUnmetRequirements) {
 
-    static Provider determineProvider(final String userAgentProvider, final List<Provider> providers) {
+        Provider result = null;
 
         if (userAgentProvider != null) {
             for (final Provider provider : providers) {
                 if (provider.getClassName().equals(userAgentProvider)) {
-                    return provider;
+                    result = provider;

```

Comment:
(Using my personal account since I don't want to do 2FA)
I personally prefer *not* to complicate this further.  What I had in mind was a simple check that the consumer calls and can get a true / false decision.  Everything else is icing on the cake - I don't think they are essential to my use case.  

I will leave it up to you to decide what kind of checks you provide, but I don't oppose just change the Javadoc to make the description accurate.  :)